cranelift/isle/veri/veri_engine/src/solver.rs

```diff
@@ -629,10 +679,35 @@ impl SolverCtx {
                         "bvnot"
                     }
                 };
-                let subexp = self.vir_expr_to_sexp(*arg);
-                self.smt.list(vec![self.smt.atom(op), subexp])
+                let sub_state: State = self.vir_expr_to_sexp(*arg);
+                let val = self.smt.list(vec![self.smt.atom(op), sub_state.val]);
+                State {
+                    val: val,
+                    ..sub_state
+                }
             }
             Expr::Binary(op, x, y) => {
+                // Look at both arg's states
```

Comment:
We might want to replace this logic with a call to the new helper for merging states, since it will simplify several cases